README.md
# Line Chart

[![pub package](https://img.shields.io/pub/v/line_chart.svg?style=for-the-badge&color=blue)](https://pub.dartlang.org/packages/line_chart)

A simple flutter package to create a custom line chart

![LineChart](https://raw.githubusercontent.com/lucas-sesti/line_chart/master/example/Showcase.jpeg)

## Installations

Add `line_chart: ^1.0.9` in your `pubspec.yaml` dependencies. And import it:

```dart
import 'package:line_chart/charts/line-chart.widget.dart';
```

## How to use

Simply create a `LineChart` widget and pass the required params:

```dart
LineChart(
  width: 300, // Width size of chart
  height: 180, // Height size of chart
  data: [], // The value to the chart
  linePaint: Paint()
      ..strokeWidth = 3
      ..style = PaintingStyle.stroke
      ..color = Colors.black, // Custom paint for the line 
),
```

Here is the widget with all options for use.

```dart
LineChart(
  width: 300, // Width size of chart
  height: 180, // Height size of chart
  data: [], // The value to the chart
  linePaint: Paint()
      ..strokeWidth = 3
      ..style = PaintingStyle.stroke
      ..color = Colors.black, // Custom paint for the line 
  circlePaint: Paint()..color = Colors.black, // Custom paint for the line 
  showPointer: true, // When press or pan update the chart, create a pointer in approximated value (The default is true)
  showCircles: true, // Show the circle in every value of chart
  customDraw: (Canvas canvas, Size size) {}, // You can draw anything in your chart, this callback is called when is generating the chart
  circleRadiusValue: 6, // The radius value of circle
  linePointerDecoration: BoxDecoration(
    color: Colors.black,
  ) // Your line pointer decoration,
  pointerDecoration: BoxDecoration(
    shape: BoxShape.circle,
    color: Colors.black,
  ) // Your decoration of circle pointer,
  insideCirclePaint: Paint()..color = Colors.white // On your circle of the chart, have a second circle, which is inside and a slightly smaller size.
  onValuePointer: (MonthChartModel value) {
    print('onValuePointer');
  } // This callback is called when change the pointer,
  onDropPointer: () {
    print('onDropPointer');
  }, // This callback is called when it is on the pointer and removes your finger from the screen
),
```

For a more detail example please take a look at the `example` folder.

## -

If something is missing, feel free to open a ticket or contribute!


CHANGELOG.md
## [1.1.11] - Line chart

- Fixed changelog and pubspec

## [1.1.10] - Line chart

- Added null safety
## [1.0.9] - Line chart

- Removed library name in painter file

## [1.0.8] - Line chart

- Fixed documentation 

## [1.0.7] - Line chart

- Removed containers in painter

## [1.0.6] - Line chart

- Created new model LineChartModelCallback when change pointer
- Now in callback of pointer, returns the percentage of amount in canvas height size

## [1.0.5] - Line chart

- Fixed name of line chart model file and link of README.md

## [1.0.4] - Line chart

- Fix bug when have many items and draw the last line

## [1.0.3] - Line chart

- Added showcase image in readme.md

## [1.0.2] - Line chart

- Removed show legends options

## [1.0.1] - Line chart

- Fix name of data model

## [1.0.0] - Line chart

- A simple line chart with full customization


test/line_chart_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:line_chart/line_chart.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
    expect(() => calculator.addOne(null), throwsNoSuchMethodError);
  });
}


lib/charts/line-chart.widget.dart
library line_chart;

import 'package:flutter/material.dart';
import 'package:line_chart/charts/line-chart-painter.widget.dart';
import 'package:line_chart/model/line-chart.model.dart';

class LineChart extends StatefulWidget {
  const LineChart({
    required this.width,
    required this.height,
    required this.data,
    required this.linePaint,
    this.circlePaint,
    this.showPointer = false,
    this.onValuePointer,
    this.onDropPointer,
    this.customDraw,
    this.insideCirclePaint,
    this.circleRadiusValue = 6,
    this.showCircles = false,
    this.linePointerDecoration,
    this.pointerDecoration,
    this.insidePadding = 8,
  });

  final Function(LineChartModelCallback)? onValuePointer;
  final Function(Canvas, Size)? customDraw;
  final Function? onDropPointer;
  final BoxDecoration? pointerDecoration;
  final BoxDecoration? linePointerDecoration;
  final double width;
  final double height;
  final double circleRadiusValue;
  final bool showPointer;
  final bool showCircles;
  final Paint linePaint;
  final Paint? circlePaint;
  final Paint? insideCirclePaint;
  final List<LineChartModel> data;
  final double insidePadding;

  @override
  _LineChartState createState() => _LineChartState();
}

class _LineChartState extends State<LineChart> {
  List<List> offsetsAndValues = [];
  BoxDecoration? linePointerDecoration = BoxDecoration(
    color: Colors.black,
  );
  BoxDecoration? pointerDecoration = BoxDecoration(
    color: Colors.black,
    shape: BoxShape.circle,
  );
  double? x = 0;
  double? y = 0;
  double? maxValue = 0;
  double? minValue = 0;
  bool showPointer = false;
  Paint? circlePaint = Paint()..color = Colors.black;
  List<double> percentagesOffsets = [];

  void initState() {
    super.initState();
    maxValue = _getMaxValue();
    minValue = _getMinValue();
    generateOffsets();

    if (widget.linePointerDecoration != null) {
      linePointerDecoration = widget.linePointerDecoration;
    }

    if (widget.pointerDecoration != null) {
      pointerDecoration = widget.pointerDecoration;
    }

    if (widget.circlePaint != null) {
      circlePaint = widget.circlePaint;
    }
  }

  void didUpdateWidget(Widget oldWidget) {
    super.didUpdateWidget(oldWidget as LineChart);

    setState(() {
      maxValue = _getMaxValue();
      minValue = _getMinValue();
      generateOffsets();

      if (widget.linePointerDecoration != null) {
        linePointerDecoration = widget.linePointerDecoration;
      }

      if (widget.pointerDecoration != null) {
        pointerDecoration = widget.pointerDecoration;
      }
    });
  }

  void generateOffsets() {
    double basePos = widget.width / widget.data.length;
    double next = 0;

    offsetsAndValues = widget.data.map((chart) {
      final Offset circlePosition =
          _getPointPos(next + widget.circleRadiusValue, chart.amount);

      next = next + basePos;
      return [
        circlePosition,
        chart,
      ];
    }).toList();
  }

  double? _getMaxValue() {
    return widget.data.fold(0, (previousValue, element) {
      if (element.amount! > previousValue!) {
        return element.amount;
      }

      return previousValue;
    });
  }

  double? _getMinValue() {
    return widget.data.fold(maxValue, (previousValue, element) {
      if (element.amount! < previousValue!) {
        return element.amount;
      }

      return previousValue;
    });
  }

  Offset _getPointPos(double width, double? amount) {
    if (maxValue == 0) {
      return Offset(width, widget.height);
    }

    double percentage = (amount! - minValue!) / (maxValue! - minValue!);

    if (percentage.isNaN) {
      percentage = 0.5;
    }

    percentagesOffsets.add(percentage);

    return Offset(
      width - widget.circleRadiusValue * 2,
      widget.height * (1 - percentage),
    );
  }

  int? _binarySearchApprox(arr, double? value) {
    if (arr.length == 1) return 0;

    int? minIndex = 0;
    var maxIndex = arr.length - 1;
    var isAscending = arr[maxIndex][0].dx > arr[minIndex][0].dx;

    while (maxIndex - minIndex > 1) {
      var curIndex = (maxIndex + minIndex) >> 1;

      if (arr[curIndex][0].dx == value) {
        return curIndex;
      }

      if (arr[curIndex][0].dx > value == isAscending) {
        maxIndex = curIndex;
      } else {
        minIndex = curIndex;
      }
    }

    if ((value! - arr[minIndex][0].dx).abs() <
        (value - arr[maxIndex][0].dx).abs()) {
      return minIndex;
    }

    return maxIndex;
  }

  void _showDetailsPointer(details) {
    if (widget.showPointer) {
      final int? nearestIndex =
          _binarySearchApprox(offsetsAndValues, details.localPosition.dx);

      setState(() {
        x = offsetsAndValues[nearestIndex!][0].dx;
        y = offsetsAndValues[nearestIndex][0].dy;

        if (details.localPosition.dx < 0) {
          showPointer = false;
        } else if (details.localPosition.dx > offsetsAndValues.last[0].dx) {
          showPointer = false;
        } else {
          if (widget.onValuePointer != null) {
            widget.onValuePointer!(
              LineChartModelCallback(
                chart: offsetsAndValues[nearestIndex][1],
                percentage: percentagesOffsets[nearestIndex] * 100,
              ),
            );
          }

          showPointer = true;
        }
      });
    }
  }

  void _dropPointer(details) {
    if (widget.showPointer) {
      setState(() => showPointer = false);

      if (widget.onDropPointer != null) {
        widget.onDropPointer!();
      }
    }
  }

  @override
  Widget build(BuildContext context) {
    return SizedBox(
      width: widget.width,
      height: widget.height + widget.insidePadding,
      child: Stack(
        alignment: Alignment.center,
        children: [
          GestureDetector(
            onLongPressStart: _showDetailsPointer,
            onLongPressMoveUpdate: _showDetailsPointer,
            onLongPressEnd: _dropPointer,
            onPanUpdate: _showDetailsPointer,
            onPanEnd: _dropPointer,
            child: CustomPaint(
              size: Size(widget.width, widget.height),
              painter: LineChartPainter(
                offsetsAndValues,
                widget.width,
                widget.height,
                widget.linePaint,
                widget.circlePaint,
                customDraw: widget.customDraw,
                insideCirclePaint: widget.insideCirclePaint,
                radiusValue: widget.circleRadiusValue,
                showCircles: widget.showCircles,
                insidePadding: widget.insidePadding,
              ),
            ),
          ),
          if (widget.showPointer) ...{
            // Line
            Positioned(
              left: x! + widget.insidePadding - 1.5,
              top: 0,
              child: AnimatedOpacity(
                duration: Duration(milliseconds: 200),
                opacity: showPointer ? 1 : 0,
                curve: Curves.easeInOut,
                child: Container(
                  height: widget.height + widget.insidePadding,
                  width: 2,
                  decoration: linePointerDecoration,
                ),
              ),
            ),

            // Circle
            Positioned(
              left: x! + widget.insidePadding - 6.5,
              top: (y! - 6) + widget.insidePadding / 2,
              child: AnimatedOpacity(
                duration: Duration(milliseconds: 200),
                opacity: showPointer ? 1 : 0,
                curve: Curves.easeInOut,
                child: Container(
                  width: widget.circleRadiusValue * 2,
                  height: widget.circleRadiusValue * 2,
                  decoration: pointerDecoration,
                ),
              ),
            ),
          },
          // if (widget.showLegend) ...{
          //   Padding(
          //     padding: EdgeInsets.only(top: 6),
          //     child: Row(
          //       children: widget.data.map<Widget>((chart) {
          //         return Expanded(
          //           child: Padding(
          //             padding: EdgeInsets.only(top: 6),
          //             child: Text('item'),
          //           ),
          //         );
          //       }).toList(),
          //     ),
          //   ),
          // }
        ],
      ),
    );
  }
}


lib/charts/time-series-chart.widget.dart
import 'package:flutter/material.dart';
import 'package:line_chart/charts/time-series-chart-painter.widget.dart';
import 'package:line_chart/model/month-bar-chart.model.dart';

class TimeSeriesChart extends StatefulWidget {
  const TimeSeriesChart({
    @required this.width,
    @required this.height,
    @required this.data,
    @required this.linePaint,
    @required this.circlePaint,
    this.showLegend = false,
    this.showPointer = false,
    this.onValuePointer,
    this.onDropPointer,
    this.customDraw,
    this.insideCirclePaint,
    this.circleRadiusValue = 6,
    this.showCircles = false,
    this.linePointerDecoration,
    this.pointerDecoration,
  });

  final double width;
  final double height;
  final List<MonthChartModel> data;
  final Paint linePaint;
  final bool showPointer;
  final Paint circlePaint;
  final BoxDecoration pointerDecoration;
  final bool showLegend;
  final Function(MonthChartModel) onValuePointer;
  final Function onDropPointer;
  final Function(Canvas, Size) customDraw;
  final Paint insideCirclePaint;
  final double circleRadiusValue;
  final bool showCircles;
  final BoxDecoration linePointerDecoration;

  @override
  _TimeSeriesChartState createState() => _TimeSeriesChartState();
}

class _TimeSeriesChartState extends State<TimeSeriesChart> {
  final double radiusValue = 6;
  BoxDecoration linePointerDecoration = BoxDecoration(
    color: Colors.black,
  );
  BoxDecoration pointerDecoration = BoxDecoration(
    color: Colors.black,
    shape: BoxShape.circle,
  );
  double x = 0;
  double y = 0;
  bool showPointer = false;
  List<List> offsetsAndValues = [];

  double maxValue = 0;
  double minValue = 0;

  void initState() {
    super.initState();
    maxValue = _getMaxValue();
    minValue = _getMinValue();
    generateOffsets();

    if (widget.linePointerDecoration != null) {
      linePointerDecoration = widget.linePointerDecoration;
    }

    if (widget.pointerDecoration != null) {
      pointerDecoration = widget.pointerDecoration;
    }
  }

  void didUpdateWidget(Widget oldWidget) {
    super.didUpdateWidget(oldWidget);

    setState(() {
      maxValue = _getMaxValue();
      minValue = _getMinValue();
      generateOffsets();

      if (widget.linePointerDecoration != null) {
        linePointerDecoration = widget.linePointerDecoration;
      }

      if (widget.pointerDecoration != null) {
        pointerDecoration = widget.pointerDecoration;
      }
    });
  }

  void generateOffsets() {
    double basePos = widget.width / widget.data.length;
    double next = 0;

    offsetsAndValues = widget.data.map((chart) {
      final Offset circlePosition = _getPointPos(next + 12, chart.amount);

      next = next + basePos;
      return [
        circlePosition,
        chart,
      ];
    }).toList();
  }

  double _getMaxValue() {
    return widget.data.fold(0, (previousValue, element) {
      if (element.amount > previousValue) {
        return element.amount;
      }

      return previousValue;
    });
  }

  double _getMinValue() {
    return widget.data.fold(maxValue, (previousValue, element) {
      if (element.amount < previousValue) {
        return element.amount;
      }

      return previousValue;
    });
  }

  Offset _getPointPos(double width, double amount) {
    if (maxValue == 0) {
      return Offset(width, widget.height);
    }

    double percentage = (amount - minValue) / (maxValue - minValue);

    if (percentage.isNaN) {
      percentage = 0.5;
    }

    return Offset(
      width - radiusValue * 2,
      widget.height * (1 - percentage),
    );
  }

  int _binarySearchApprox(arr, double value) {
    if (arr.length == 1) return 0;

    var minIndex = 0;
    var maxIndex = arr.length - 1;
    var isAscending = arr[maxIndex][0].dx > arr[minIndex][0].dx;

    while (maxIndex - minIndex > 1) {
      var curIndex = (maxIndex + minIndex) >> 1;

      if (arr[curIndex][0].dx == value) {
        return curIndex;
      }

      if (arr[curIndex][0].dx > value == isAscending) {
        maxIndex = curIndex;
      } else {
        minIndex = curIndex;
      }
    }

    if ((value - arr[minIndex][0].dx).abs() <
        (value - arr[maxIndex][0].dx).abs()) {
      return minIndex;
    }

    return maxIndex;
  }

  void _showDetailsPointer(details) {
    if (widget.showPointer) {
      final int nearestIndex =
          _binarySearchApprox(offsetsAndValues, details.localPosition.dx);

      setState(() {
        x = offsetsAndValues[nearestIndex][0].dx;
        y = offsetsAndValues[nearestIndex][0].dy;

        if (details.localPosition.dx < 0) {
          showPointer = false;
        } else if (details.localPosition.dx > offsetsAndValues.last[0].dx) {
          showPointer = false;
        } else {
          if (widget.onValuePointer != null) {
            widget.onValuePointer(offsetsAndValues[nearestIndex][1]);
          }

          showPointer = true;
        }
      });
    }
  }

  void _dropPointer(details) {
    if (widget.showPointer) {
      setState(() => showPointer = false);

      if (widget.onDropPointer != null) {
        widget.onDropPointer();
      }
    }
  }

  @override
  Widget build(BuildContext context) {
    return SizedBox(
      width: widget.width,
      height: widget.height + 16,
      child: Stack(
        alignment: Alignment.center,
        children: [
          Padding(
            padding: EdgeInsets.symmetric(
              horizontal: 16,
            ),
            child: GestureDetector(
              onLongPressStart: _showDetailsPointer,
              onLongPressMoveUpdate: _showDetailsPointer,
              onLongPressEnd: _dropPointer,
              onPanUpdate: _showDetailsPointer,
              onPanEnd: _dropPointer,
              child: CustomPaint(
                size: Size(widget.width, widget.height),
                painter: TimeSeriesChartPainter(offsetsAndValues, widget.width,
                    widget.height, widget.linePaint, widget.circlePaint,
                    customDraw: widget.customDraw,
                    insideCirclePaint: widget.insideCirclePaint,
                    radiusValue: widget.circleRadiusValue,
                    showCircles: widget.showCircles),
              ),
            ),
          ),
          if (widget.showPointer) ...{
            // Line
            Positioned(
              left: x + 14.5,
              top: 0,
              child: AnimatedOpacity(
                duration: Duration(milliseconds: 200),
                opacity: showPointer ? 1 : 0,
                curve: Curves.easeInOut,
                child: Container(
                  height: widget.height + 16,
                  width: 2,
                  decoration: linePointerDecoration,
                ),
              ),
            ),

            // Circle
            Positioned(
              left: x + 9,
              top: y + 2,
              child: AnimatedOpacity(
                duration: Duration(milliseconds: 200),
                opacity: showPointer ? 1 : 0,
                curve: Curves.easeInOut,
                child: Container(
                  width: 12,
                  height: 12,
                  decoration: pointerDecoration,
                ),
              ),
            ),
          },
          if (widget.showLegend) ...{
            Padding(
              padding: EdgeInsets.only(top: 6),
              child: Row(
                children: widget.data.map<Widget>((chart) {
                  return Expanded(
                    child: Padding(
                      padding: EdgeInsets.only(top: 6),
                      child: Text('item'),
                    ),
                  );
                }).toList(),
              ),
            ),
          }
        ],
      ),
    );
  }
}


lib/charts/line-chart-painter.widget.dart
import 'package:flutter/material.dart';

class LineChartPainter extends CustomPainter {
  final List<List> data;
  final double width;
  final double height;
  final Function(Canvas, Size)? customDraw;
  final Paint linePaint;
  final Paint? circlePaint;
  final bool showCircles;
  final double radiusValue;
  final Paint? insideCirclePaint;
  final double insidePadding;

  LineChartPainter(
    this.data,
    this.width,
    this.height,
    this.linePaint,
    this.circlePaint, {
    this.customDraw,
    this.showCircles = true,
    this.radiusValue = 6,
    this.insideCirclePaint,
    required this.insidePadding,
  });

  @override
  void paint(Canvas canvas, Size size) {
    if (customDraw != null) {
      customDraw!(canvas, size);
    }

    data.forEach((value) {
      final currentIndex = data.indexOf(value);

      Path path = Path();

      path.moveTo(value[0].dx + insidePadding, value[0].dy);

      if (data.last != value) {
        path.lineTo(data[currentIndex + 1][0].dx + insidePadding + 1,
            data[currentIndex + 1][0].dy);
      }

      canvas.drawPath(path, linePaint);

      if (showCircles) {
        canvas.drawCircle(Offset(value[0].dx + insidePadding, value[0].dy),
            radiusValue, circlePaint!);
        canvas.drawCircle(
          Offset(value[0].dx + insidePadding, value[0].dy),
          radiusValue - radiusValue / 2,
          insideCirclePaint != null ? insideCirclePaint! : circlePaint!,
        );
      }
    });
  }

  @override
  bool shouldRepaint(CustomPainter old) {
    return true;
  }
}


lib/charts/time-series-chart-painter.widget.dart
import 'package:flutter/material.dart';

class TimeSeriesChartPainter extends CustomPainter {
  final List<List> data;
  final double width;
  final double height;
  final Function(Canvas, Size) customDraw;
  final Paint linePaint;
  final Paint circlePaint;
  final bool showCircles;
  final double radiusValue;
  final Paint insideCirclePaint;

  TimeSeriesChartPainter(
    this.data,
    this.width,
    this.height,
    this.linePaint,
    this.circlePaint, {
    this.customDraw,
    this.showCircles = true,
    this.radiusValue = 6,
    this.insideCirclePaint,
  });

  @override
  void paint(Canvas canvas, Size size) {
    if (customDraw != null) {
      customDraw(canvas, size);
    }

    data.forEach((value) {
      final currentIndex = data.indexOf(value);

      Path path = Path();

      path.moveTo(value[0].dx, value[0].dy);

      if (data.last == value) {
        path.lineTo(size.width, value[0].dy);
      } else {
        path.lineTo(data[currentIndex + 1][0].dx, data[currentIndex + 1][0].dy);
      }

      canvas.drawPath(path, linePaint);

      if (showCircles) {
        canvas.drawCircle(value[0], radiusValue, circlePaint);
        canvas.drawCircle(
          value[0],
          radiusValue - radiusValue / 2,
          insideCirclePaint != null ? insideCirclePaint : circlePaint,
        );
      }
    });
  }

  @override
  bool shouldRepaint(CustomPainter old) {
    return true;
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:line_chart/charts/line-chart.widget.dart';
import 'package:line_chart/model/line-chart.model.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blue,
        // This makes the visual density adapt to the platform that you run
        // the app on. For desktop platforms, the controls will be smaller and
        // closer together (more dense) than on mobile platforms.
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key, this.title}) : super(key: key);

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String? title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  List<LineChartModel> data = [
    LineChartModel(amount: 300, date: DateTime(2020, 1, 1)),
    LineChartModel(amount: 200, date: DateTime(2020, 1, 2)),
    LineChartModel(amount: 300, date: DateTime(2020, 1, 3)),
    LineChartModel(amount: 500, date: DateTime(2020, 1, 4)),
    LineChartModel(amount: 800, date: DateTime(2020, 1, 5)),
    LineChartModel(amount: 200, date: DateTime(2020, 1, 6)),
    LineChartModel(amount: 120, date: DateTime(2020, 1, 7)),
    LineChartModel(amount: 140, date: DateTime(2020, 1, 8)),
    LineChartModel(amount: 110, date: DateTime(2020, 1, 9)),
    LineChartModel(amount: 250, date: DateTime(2020, 1, 10)),
    LineChartModel(amount: 390, date: DateTime(2020, 1, 11)),
    LineChartModel(amount: 1300, date: DateTime(2020, 1, 12)),
  ];

  int _counter = 0;

  void _incrementCounter() {
    setState(() {
      data = [
        LineChartModel(amount: 500, date: DateTime(2020, 1, 4)),
        LineChartModel(amount: 200, date: DateTime(2020, 1, 2)),
        LineChartModel(amount: 200, date: DateTime(2020, 1, 6)),
        LineChartModel(amount: 800, date: DateTime(2020, 1, 5)),
        LineChartModel(amount: 1300, date: DateTime(2020, 1, 12)),
        LineChartModel(amount: 300, date: DateTime(2020, 1, 3)),
        LineChartModel(amount: 120, date: DateTime(2020, 1, 7)),
        LineChartModel(amount: 250, date: DateTime(2020, 1, 10)),
        LineChartModel(amount: 140, date: DateTime(2020, 1, 8)),
        LineChartModel(amount: 100, date: DateTime(2020, 1, 1)),
        LineChartModel(amount: 390, date: DateTime(2020, 1, 11)),
        LineChartModel(amount: 110, date: DateTime(2020, 1, 9)),
        LineChartModel(amount: 410, date: DateTime(2020, 4, 9)),
      ];
    });
  }

  @override
  Widget build(BuildContext context) {
    Paint circlePaint = Paint()..color = Colors.black;

    Paint insideCirclePaint = Paint()..color = Colors.white;

    Paint linePaint = Paint()
      ..strokeWidth = 3
      ..style = PaintingStyle.stroke
      ..color = Colors.black;

    return Scaffold(
      appBar: AppBar(
        title: Text('Line Chart Showcase'),
      ),
      body: ListView(
        children: [
          // First chart
          Column(
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              Text(
                'Line chart - Line, Circles, Pointer',
                style: Theme.of(context).textTheme.headline5,
              ),
              LineChart(
                width: MediaQuery.of(context).size.width,
                height: 180,
                data: data,
                linePaint: linePaint,
                circlePaint: circlePaint,
                showPointer: true,
                showCircles: true,
                customDraw: (Canvas canvas, Size size) {},
                linePointerDecoration: BoxDecoration(
                  color: Colors.black,
                ),
                pointerDecoration: BoxDecoration(
                  shape: BoxShape.circle,
                  color: Colors.black,
                ),
                insideCirclePaint: insideCirclePaint,
                onValuePointer: (LineChartModelCallback value) {
                  print('${value.chart} ${value.percentage}');
                },
                onDropPointer: () {
                  print('onDropPointer');
                },
                insidePadding: 16,
              ),
            ],
          ),

          // Second Chart
          Padding(
            padding: EdgeInsets.symmetric(
              vertical: 24,
            ),
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                Text(
                  'Line chart - Line, Pointer',
                  style: Theme.of(context).textTheme.headline5,
                ),
                LineChart(
                  width: MediaQuery.of(context).size.width,
                  height: 180,
                  data: data,
                  linePaint: linePaint,
                  circlePaint: circlePaint,
                  showPointer: true,
                  showCircles: false,
                  customDraw: (Canvas canvas, Size size) {},
                  linePointerDecoration: BoxDecoration(
                    color: Colors.black,
                  ),
                  pointerDecoration: BoxDecoration(
                    shape: BoxShape.circle,
                    color: Colors.black,
                  ),
                  insideCirclePaint: insideCirclePaint,
                  onValuePointer: (LineChartModelCallback value) {
                    print('${value.chart} ${value.percentage}');
                  },
                  onDropPointer: () {
                    print('onDropPointer');
                  },
                ),
              ],
            ),
          ),

          // Third chart
          Column(
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              Text(
                'Line chart - Line',
                style: Theme.of(context).textTheme.headline5,
              ),
              LineChart(
                width: MediaQuery.of(context).size.width,
                height: 180,
                data: data,
                linePaint: linePaint,
                circlePaint: circlePaint,
                showPointer: false,
                showCircles: false,
                customDraw: (Canvas canvas, Size size) {},
                linePointerDecoration: BoxDecoration(
                  color: Colors.black,
                ),
                pointerDecoration: BoxDecoration(
                  shape: BoxShape.circle,
                  color: Colors.black,
                ),
                insideCirclePaint: insideCirclePaint,
                onValuePointer: (LineChartModelCallback value) {
                  print('${value.chart} ${value.percentage}');
                },
                onDropPointer: () {
                  print('onDropPointer');
                },
              ),
            ],
          ),
        ],
      ),
    );
  }
}
